sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -1374,33 +1374,110 @@ case class Pivot(
+ * Either ids or values array must be set. The ids array can be empty,
+ * the values array must not be empty if not None.
+ *
+ * A None ids array will be replaced during analysis with all resolved outputs of child except
+ * the values. This expansion allows to easily select all non-value columns as id columns.
+ *
+ * A None values array will be replaced during analysis with all resolved outputs of child except
  * the ids. This expansion allows to easily unpivot all non-id columns.
  *
  * @see `org.apache.spark.sql.catalyst.analysis.Analyzer.ResolveUnpivot`
  *
- * The type of the value column is derived from all value columns during analysis once all values
- * are resolved. All values' types have to be compatible, otherwise the result value column cannot
- * be assigned the individual values and an AnalysisException is thrown.
+ * Multiple columns can be unpivoted in one row by providing multiple value column names
+ * and the same number of unpivot value expressions:
+ * {{{
+ *   // one-dimensional value columns
+ *   Unpivot(
+ *     Some(Seq("id")),
+ *     Some(Seq(
+ *       (Seq("val1"), None),
+ *       (Seq("val2"), None)
+ *     )),
+ *     "var",
+ *     Seq("val")
+ *   )
+ *
+ *   // two-dimensional value columns
+ *   Unpivot(
+ *     Some(Seq("id")),
+ *     Some(Seq(
+ *       (Seq("val1.1", "val1.2"), None),
+ *       (Seq("val2.1", "val2.2"), None)
+ *     )),
+ *     "var",
+ *     Seq("val1", "val2")
+ *   )
+ * }}}
+ *
+ * The variable column will contain the name of the unpivot value while the value columns contain
+ * the unpivot values. Multi-dimensional unpivot values can be given `aliases`:
+ * }}}
+ *   // two-dimensional value columns with aliases
+ *   Unpivot(
+ *     Some(Seq("id")),
+ *     Some(Seq(
+ *       (Seq("val1.1", "val1.2"), Some("val1")),
+ *       (Seq("val2.1", "val2.2"), Some("val2"))
+ *     )),
+ *     "var",
+ *     Seq("val1", "val2")
+ *   )
+ * }}}
+ *
+ * All "value" columns must share a least common data type. Unless they are the same data type,
+ * all "value" columns are cast to the nearest common data type. For instance,
+ * types `IntegerType` and `LongType` are cast to `LongType`, while `IntegerType` and `StringType`
+ * do not have a common data type and `unpivot` fails with an `AnalysisException`.
  *
  * @see `org.apache.spark.sql.catalyst.analysis.TypeCoercionBase.UnpivotCoercion`
  *
  * @param ids                Id columns
- * @param values             Value columns to unpivot
+ * @param values             Value column sets to unpivot with optional aliases
  * @param variableColumnName Name of the variable column
- * @param valueColumnName    Name of the value column
+ * @param valueColumnNames   Names of the value columns
  * @param child              Child operator
  */
 case class Unpivot(
-    ids: Seq[NamedExpression],
-    values: Seq[NamedExpression],
+    ids: Option[Seq[NamedExpression]],
+    values: Option[Seq[UnpivotExpr]],
```

Comment:
Sorry for the back and forth, but after reading the source code more, the alias problem is a bit hard to solve. Today Spark has 2 frameworks to enforce the alias:
1. allow a list of operators to hold `NamedExpression`s, such as `Project` and `Aggregate`. The rule `CleanupAliases` handles these operators specially to not remove top-level aliases, to keep `NamedExpression`
2. allow a list of expressions to hold alias names, such as `ArraysZip` and `CreateStruct`. These expressions will contain `NamePlaceholder` first, and then the rule `ResolveExpressionsWithNamePlaceholders` will match these expressions and resolve `NamePlaceholder`.

`Unpivot` is an operator and it fits the first framework better. So it seems simpler to use `values: Option[Seq[Seq[NamedExpression]]]` with an extra `aliases: Seq[Option[String]]` parameter.

Reply:
or we can do `values: Option[Seq[(Seq[NamedExpression], Option[String])]]` and update `QueryPlan.mapExpressions` to support `Tuple2`.

Reply:
I'd prefer the tuple, but it makes the code harder to read, so I revert back to the `aliases` field: cf0ec27